File: Tests/GlasstualTests/IRCAddressBookTests.m
/* *********************************************************************
 * Copyright (c) 2026 Codeux Software, LLC & respective contributors.
 * Please see Acknowledgements.pdf for additional information.
 *********************************************************************** */

#import <XCTest/XCTest.h>

#import "IRCAddressBook.h"

NS_ASSUME_NONNULL_BEGIN

@interface IRCAddressBookTests : XCTestCase
@end

@implementation IRCAddressBookTests

- (void)testIgnoreEntryTreatsRegularExpressionCharactersLiterally
{
	IRCAddressBookEntry *entry = [IRCAddressBookEntry newIgnoreEntryForHostmask:@"nick[1]!*@example.com"];

	XCTAssertTrue([entry checkMatch:@"NICK[1]!user@example.com"]);
	XCTAssertFalse([entry checkMatch:@"nick1!user@example.com"]);
	XCTAssertEqualObjects(entry.hostmaskRegularExpression, @"^nick\\[1]!.*?@example\\.com$");
}

- (void)testIgnoreEntrySupportsIRCWildcardsAndAnchorsTheMatch
{
	IRCAddressBookEntry *entry = [IRCAddressBookEntry newIgnoreEntryForHostmask:@"n?ck!*@*.example"];

	XCTAssertTrue([entry checkMatch:@"nick!user@irc.example"]);
	XCTAssertFalse([entry checkMatch:@"prefix-nick!user@irc.example"]);
	XCTAssertFalse([entry checkMatch:@"noock!user@irc.example"]);
}

- (void)testUserTrackingEntryDerivesNicknameAndMatchesFullHostmask
{
	IRCAddressBookEntryMutable *entry = [IRCAddressBookEntryMutable newUserTrackingEntry];

	entry.hostmask = @"Alice";

	XCTAssertEqualObjects(entry.trackingNickname, @"Alice");
	XCTAssertTrue([entry checkMatch:@"alice!user@example.com"]);
	XCTAssertFalse([entry checkMatch:@"alice"]);
}

- (void)testMixedEntryHasNoMatcherState
{
	IRCAddressBookEntryMutable *entry = [IRCAddressBookEntryMutable newIgnoreEntry];

	entry.entryType = IRCAddressBookEntryTypeMixed;

	XCTAssertEqualObjects(entry.hostmaskRegularExpression, @"");
	XCTAssertNil(entry.trackingNickname);
	XCTAssertFalse([entry checkMatch:@""]);
}

@end

NS_ASSUME_NONNULL_END

File: Tests/GlasstualTests/IRCAddressBookTests.swift
@testable import Glasstual
import Testing

/** *********************************************************************
 * Copyright (c) 2026 Codeux Software, LLC & respective contributors.
 * Please see Acknowledgements.pdf for additional information.
 *********************************************************************** */
@MainActor
@Suite("Address book matching")
struct IRCAddressBookTests {
	@Test("A regular expression character in a hostmask matches itself")
	func ignoreEntryTreatsRegularExpressionCharactersLiterally() {
		let entry = AddressBookEntry
			.newIgnoreEntry(forHostmask: "nick[1]!*@example.com")

		#expect(entry.checkMatch("NICK[1]!user@example.com"))
		#expect(entry.checkMatch("nick1!user@example.com") == false)
		#expect(entry.hostmaskRegularExpression == #"^nick\[1]!.*?@example\.com$"#)
	}

	@Test("IRC wildcards are honoured and the match is anchored at both ends")
	func ignoreEntrySupportsIRCWildcardsAndAnchorsTheMatch() {
		let entry = AddressBookEntry
			.newIgnoreEntry(forHostmask: "n?ck!*@*.example")

		#expect(entry.checkMatch("nick!user@irc.example"))
		#expect(entry.checkMatch("prefix-nick!user@irc.example") == false)
		#expect(entry.checkMatch("noock!user@irc.example") == false)
	}

	@Test("A tracking entry takes its nickname from the hostmask and matches the whole mask")
	func userTrackingEntryDerivesNicknameAndMatchesFullHostmask() {
		var entry = AddressBookEntry.newUserTrackingEntry()

		entry.hostmask = "Alice"

		#expect(entry.trackingNickname == "Alice")
		#expect(entry.checkMatch("alice!user@example.com"))
		#expect(entry.checkMatch("alice") == false)
	}

	@Test("A mixed entry carries no matcher state at all")
	func mixedEntryHasNoMatcherState() {
		var entry = AddressBookEntry.newIgnoreEntry()

		entry.entryType = .mixed

		#expect(entry.hostmaskRegularExpression == "")
		#expect(entry.trackingNickname == nil)
		#expect(entry.checkMatch("") == false)
	}
}
